Mathlib/GroupTheory/SpecificGroups/KleinFour.lean

```diff
@@ -39,74 +40,195 @@ produces the third one.
+def instCommGroupOfExponentTwo (hG : Monoid.exponent G = 2) : CommGroup G where
+  mul_comm := mul_comm_of_exponent_two hG
+
+@[to_additive]
+lemma mul_not_mem_of_orderOf_eq_two {G : Type*} [Group G] {x y : G} (hx : orderOf x = 2)
+    (hy : orderOf y = 2) (hxy : x ≠ y) : x * y ∉ ({x, y, 1} : Set G) := by
+  simp only [Set.mem_singleton_iff, Set.mem_insert_iff, mul_right_eq_self, mul_left_eq_self, not_or]
+  refine ⟨?_, ?_, fun h' ↦ hxy <| by
+    simpa only [mul_eq_one_iff_eq_inv, inv_eq_self_of_orderOf_eq_two hx,
+      inv_eq_self_of_orderOf_eq_two hy] using h'⟩
+  all_goals rw [← orderOf_eq_one_iff]; norm_num [hy, hx]
+
+@[to_additive]
+lemma mul_not_mem_of_exponent_two {G : Type*} [Group G] (h : Monoid.exponent G = 2) {x y : G}
+    (hx : x ≠ 1) (hy : y ≠ 1) (hxy : x ≠ y) : x * y ∉ ({x, y, 1} : Set G) :=
+  mul_not_mem_of_orderOf_eq_two (orderOf_eq_prime (h ▸ Monoid.pow_exponent_eq_one x) hx)
+    (orderOf_eq_prime (h ▸ Monoid.pow_exponent_eq_one y) hy) hxy
+
+end ExponentTwo
+
+/-! # Klein four-groups as a mixin class -/
+
+/-- An (additive) Klein four-group is an (additive) group of cardinality four and exponent two. -/
+class IsAddKleinFour (G : Type*) [AddGroup G] : Prop where
+  card_four : Nat.card G = 4
+  exponent_two : AddMonoid.exponent G = 2
+
+/-- A Klein four-group is a group of cardinality four and exponent two. -/
+@[to_additive existing IsAddKleinFour]
+class IsKleinFour (G : Type*) [Group G] : Prop where
+  card_four : Nat.card G = 4
+  exponent_two : Monoid.exponent G = 2
+
+attribute [simp] IsKleinFour.card_four IsKleinFour.exponent_two
+  IsAddKleinFour.card_four IsAddKleinFour.exponent_two
+
+instance : IsAddKleinFour (ZMod 2 × ZMod 2) where
+  card_four := by simp
+  exponent_two := by simp [AddMonoid.exponent_prod]
+
+instance : IsKleinFour (DihedralGroup 2) where
+  card_four := by simp only [Nat.card_eq_fintype_card]; rfl
+  exponent_two := by simp [DihedralGroup.exponent]
+
+instance {G : Type*} [Group G] [IsKleinFour G] :
+    IsAddKleinFour (Additive G) where
+  card_four := by rw [← IsKleinFour.card_four (G := G)]; congr!
+  exponent_two := by simp
+
+instance {G : Type*} [AddGroup G] [IsAddKleinFour G] :
+    IsKleinFour (Multiplicative G) where
+  card_four := by rw [← IsAddKleinFour.card_four (G := G)]; congr!
+  exponent_two := by simp
+
+namespace IsKleinFour
+
+@[to_additive]
+instance instFinite {G : Type*} [Group G] [IsKleinFour G] : Finite G :=
+  Nat.finite_of_card_ne_zero <| by norm_num [IsKleinFour.card_four]
+
+@[to_additive (attr := simp)]
+lemma card_four' {G : Type*} [Group G] [Fintype G] [IsKleinFour G] :
+    Fintype.card G = 4 :=
+  Nat.card_eq_fintype_card (α := G).symm ▸ IsKleinFour.card_four
+
+open Finset
+
+variable {G : Type*} [Group G]
+
+/-- A group of order 4 is not cyclic if and only if its exponent is two. -/
+@[to_additive]
+lemma not_isCyclic_iff (hG : Nat.card G = 4) :
+    ¬ IsCyclic G ↔ Monoid.exponent G = 2 := by
```

Comment:
This statement is a special case of
```
lemma not_isCyclic_iff {p : ℕ} (hp : Prime p) (hG : Nat.card G = p ^ 2) :
    ¬ IsCyclic G ↔ Monoid.exponent G = p := by
```
Does it make sense to prove the more general one and deduce the current one from the general one?

Reply:
Sure. I can do that.

Reply:
This also holds with `p ^ n` (where `n > 1`). Maybe this can also be stated for a `PGroup` (but if it becomes too complicated let's keep this version for the PR).

Reply:
Riccardo, I am not sure that I understand your comment: ℤ/4+ℤ/2 is not cyclic and does not have exponent 2.

Reply:
It's too early in the morning, sorry.

Reply:
I was thinking about the fact that the group is cyclic iff the exponent is the cardinality of the group.

Reply:
@adomani I ended up splitting this off into #8723